creusot/src/ctx.rs

```diff
@@ -318,6 +322,168 @@ impl<'tcx> TranslationCtx<'tcx> {
+                let program_subst = erased_identity_for_item(tcx, def_id);
+                let logic_subst = erased_identity_for_item(tcx, logic_id);
+                let mut program_subst = program_subst.iter().map(|arg| arg.kind());
+                let mut logic_subst = logic_subst.iter().map(|arg| arg.kind());
+                loop {
+                    let prog_arg = program_subst.next();
+                    let logic_arg = logic_subst.next();
+                    match (prog_arg, logic_arg) {
+                        (None, None) => break,
+                        (Some(arg), None) | (None, Some(arg)) => {
+                            tcx.dcx()
+                                    .struct_span_err(
+                                        span,
+                                        format!(
+                                            "mismatched generic parameters for logical alias: {} parameter {arg:?} in the alias",
+                                            if prog_arg.is_some() {
+                                                "missing"
+                                            } else {
+                                                "additional"
+                                            }),
+                                    )
+                                    .with_span_note(program_span, "function defined here")
+                                    .with_span_note(logic_span, "logical alias defined here")
+                                    .emit();
+                        }
+                        (Some(GenericArgKind::Type(t1)), Some(GenericArgKind::Type(t2))) => {
+                            if t1 != t2 {
+                                tcx.dcx().struct_span_err(span, format!("mismatched types in logical alias: expected {t1}, got {t2}"))
+                                        .with_span_note(program_span, "function defined here")
+                                        .with_span_note(logic_span, "logical alias defined here")
+                                        .emit();
+                            }
+                        }
+                        (Some(GenericArgKind::Const(c1)), Some(GenericArgKind::Const(c2))) => {
+                            let (t1, name1) = match c1.kind() {
+                                rustc_type_ir::ConstKind::Param(p) => {
+                                    (p.find_const_ty_from_env(tcx.param_env(def_id)), p.name)
+                                }
+                                _ => unreachable!(),
+                            };
+                            let (t2, name2) = match c2.kind() {
+                                rustc_type_ir::ConstKind::Param(p) => {
+                                    (p.find_const_ty_from_env(tcx.param_env(logic_id)), p.name)
+                                }
+                                _ => unreachable!(),
+                            };
+                            if t1 != t2 {
```

Comment:
Why don't we check the equality of constant param names, too? It seems weird to do that for types and lifetimes, but not for constants.